### DISCUSSION TITLE
Handling Duplicate Messages in libp2p with gossipsub

### DISCUSSION THREAD

**Discussion by vinay10949** [libp2p/rust-libp2p#5278] [Q&A] [answered]
**Description:**

I'm using libp2p with gossipsub for message publication and subscription in my application. I have configured gossipsub with a duplicate cache time of 1 second, intending to allow duplicate messages after this interval. However, I'm encountering a "Duplicate" error when publishing messages even after the specified time interval has passed.

**Code Snippet:**

```rust
let gossipsub_config = gossipsub::ConfigBuilder::default()
    .heartbeat_interval(Duration::from_secs(10))
    .validation_mode(gossipsub::ValidationMode::Strict)
    .duplicate_cache_time(Duration::from_secs(1))
    .published_message_ids_cache_time(Duration::from_secs(1))
    .message_id_fn(message_id_fn)
    .build()
    .map_err(|msg| io::Error::new(io::ErrorKind::Other, msg))?;
```

**Questions:**

1. What settings or configurations do I need to adjust to properly handle duplicate messages after the specified time interval?
2. Are there any additional considerations or configurations related to duplicate message handling in libp2p with gossipsub that I might be missing?
3. Also How do I need to make any specific changes for gossip sub or anything in Kademlia Behaviour to allow public ipaddress for communication.

**Comment by jxs** [comment, marked as answer]
Hi, have you seen https://github.com/libp2p/rust-libp2p/issues/2587#issuecomment-1078900995 ?
But basically, when publishing via `gossipsub`, it checks on `DuplicateCache` if contains the entry:
https://github.com/libp2p/rust-libp2p/blob/ab4f664e0437ce486b0aa3401f6eccaf3bcdb541/protocols/gossipsub/src/behaviour.rs#L607-L616

but `DuplicateCache` only removes stale entries when you insert a new message:
https://github.com/libp2p/rust-libp2p/blob/ab4f664e0437ce486b0aa3401f6eccaf3bcdb541/protocols/gossipsub/src/time_cache.rs#L172-L179
https://github.com/libp2p/rust-libp2p/blob/ab4f664e0437ce486b0aa3401f6eccaf3bcdb541/protocols/gossipsub/src/time_cache.rs#L133-L144
not when `contains()` is called:
https://github.com/libp2p/rust-libp2p/blob/ab4f664e0437ce486b0aa3401f6eccaf3bcdb541/protocols/gossipsub/src/time_cache.rs#L181-L184

I'd say you are probably trying to send the same message sequentially

**Comment by AgeManning** [comment]
> What settings or configurations do I need to adjust to properly handle duplicate messages after the specified time interval?

Duplicates on gossipsub should be expressly forbidden. In an ideal world, gossipsub should never permit duplicates, however we are limited in resources (i.e RAM) such that we can't remember all messages for an infinite amount of time. Given these constraints, sometimes a duplicate can get through, in Rust, we loosely time-bound the message. The time-bound ensures that no duplicates will happen within that time (i.e 1 second) but may prevent duplicates longer. With unlimited resources, we'd set this time limit to infinity. 

This begs the question, why do you want duplicates? Duplicates are a source of amplification and excess bandwidth use in gossipsub. We don't want old messages permanently being slowly bouncing around the network and re-transmitted forever, so having a large enough cache prevents this. 

A message is determined by its message-id. In your code-snippet you have changed the default message to something else but have not said what it is. The default message-id has a sequence number associated with it. So if you had a chat app, and wanted to send the message "hey" and then send "hey" again immediately after, this would be fine. These messages are not duplicates because the message-id will be different. If you have a message whose contents need to be re-transmitted regularly, it sounds like your message-id is not correctly built for it. 

> Are there any additional considerations or configurations related to duplicate message handling in libp2p with gossipsub that I might be missing?

Yep, sounds like you need to change your message-id so that it is unique per message you want to transmit. 

> Also How do I need to make any specific changes for gossip sub or anything in Kademlia Behaviour to allow public ipaddress for communication.

I don't really follow this. Gossipsub works fine with public addresses. This is part of libp2p, you can dial any public multiaddr. I think there is minimal difference in libp2p between public/private addresses.

**Comment by dariusc93** [comment]
To add to above, the default `message_id_fn` does generate a unique message id

https://github.com/libp2p/rust-libp2p/blob/b890b2962018c86dc68bed38ea0abf7799d91406/protocols/gossipsub/src/config.rs#L389-L402

However, if youre going based on the chat example the message id would likely be the same if the message is the same because it generates it from the hash

https://github.com/libp2p/rust-libp2p/blob/b890b2962018c86dc68bed38ea0abf7799d91406/examples/chat/src/main.rs#L55-L59

I would suggest omitting that segment of code as well as the `message_id_fn` in the chain and see if that helps (though this is all assuming you are using the example. If you are using your own code to generate the id, you might want to factor in `Message::sequence_number` or some timestamp, though the default should be more than enough unless you have a use-case)
